### PR TITLE
feat(bloc_test): add `setUp` and `tearDown`

### PR DESCRIPTION
**Pull request by felangel**
## Status

**READY**

## Breaking Changes

NO

## Description

- feat(bloc_test): add `setUp` and `tearDown` methods

As a developer, I want to have the ability to execute set up and tear down code as part of a `blocTest` so that I don't have to clutter the `build` and `verify` api:

```dart
blocTest(
  setUp: () {
    // Execute any set up code needed prior to building the bloc.
    when(() => repository.someMethod()).thenReturn(...);
  },
  build: () => MyBloc(repository),
  act: (bloc) => bloc.add(...),
  expect: () => [...],
  tearDown: () {
    // Execute any tear down code needed after the test has run.
    verifyNoMoreInteractions(repository);
  },
);
```

## Type of Change

- [X] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/bloc_test/test/bloc_bloc_test_test.dart [hunk 1]
```diff
@@ -550,4 +561,30 @@ Alternatively, consider using Matchers in the expect of the blocTest rather than
...
+    tearDown(() {
+      expect(tearDownCallCount, equals(1));
+    });
+
+    blocTest<CounterBloc, int>(
+      'is called after the test is run',
+      build: () => CounterBloc(),
+      act: (bloc) => bloc.add(CounterEvent.increment),
+      verify: (bloc) {
+        state = bloc.state;
```
Comment (korzonkiee) [inline packages/bloc_test/test/bloc_bloc_test_test.dart:582]: Is it necessary to set the `state` variable here and then expect that it's correct in `blocTest.tearDown`? Maybe I'm missing something, but it doesn't seem related to testing the `tearDown` callback.

Comment (felangel) [reply packages/bloc_test/test/bloc_bloc_test_test.dart:582]: the goal of this was to ensure that tearDown executes last. This isn't how you'd use `tearDown` normally -- it's more to ensure that tearDown is called last.

Comment (korzonkiee) [reply packages/bloc_test/test/bloc_bloc_test_test.dart:582]: ok, that makes sense - thanks for claryfing.

**Review by korzonkiee — APPROVED** [review]
LGTM 👍

**Review by jorgecoca — APPROVED** [review]
I understand why these changes are suggested and its benefits, but I question I can see being raised is why/when to use these new two APIs under `blocTest`, and when to use the `setUp` and `tearDown` methods from the `test` package.

Should we add more documentation for that?

**Comment by felangel** [comment]
> I understand why these changes are suggested and its benefits, but I question I can see being raised is why/when to use these new two APIs under `blocTest`, and when to use the `setUp` and `tearDown` methods from the `test` package.
> 
> Should we add more documentation for that?

In my opinion the setUp and tearDown in the group is for common setUp and tearDown code whereas the setUp and tearDown apis in blocTest are intended to be test-specific (e.g. stubbing repository responses, etc.). These are usually test-specific and currently developers have to include that logic in the `build` api which clutters it.
